Complete Project

Kwaku28 opened the pull request:
## This project was implemented to create a Recipe App

### Features of the app include:
1. A login page.
2. A registration page.
3. A food list
4. A recipes list
5. A general shopping list view (all missing food for all your recipes and total price).
6. A list of all public recipes from other users with their names and total prices.

### Tasks completed include the following:

- [x] Login page and registration page:
     - Built with Devise.
- [x] Food list:
     - Displays a list of food added by the logged-in user(displays also the quantity of a given food).
     - Added a form that allows users to add new food.
- [x] Recipes list:
     - Displays a list of recipes created by the logged-in user
     - Lead to recipe details.
     - Allows the owner of the recipe to delete it.
- [x] Public recipe list:
     - Displays a list of all public recipes ordered by the newest.
     - Leads to recipe details.
- [x] Recipe details:
     - Displays a toggle button that allows for a recipe to be made public or private.
     - If the recipe is public or the user is the owner of the recipe, it displays the recipe details.
     - If the user is the owner of the recipe, it leads to the form that allows the user to add new food.
- [x] General shopping list view:
     - Shows the list of food that is missing for all recipes of the logged-in user (compares the list of food for all recipes with the general food list of that user).
     - Counts the total food items and total price of the missing food.
- [x] Fix N+1 queries.
- [x] Added a navigation menu allowing users to open all the pages created.
- [x] Added unit and integration tests.

Thank you🤝

touredri committed:
setup repo

Kwaku28 committed:
Configure and install linters

Kwaku28 committed:
Add rspec gem and install bundle

touredri committed:
added dependences

touredri committed:
Merge branch 'setup' of https://github.com/touredri/Recipe-app into setup

touredri committed:
setup database.yml

Kwaku28 committed:
Update readme file

touredri committed:
Merge branch 'setup' of https://github.com/touredri/Recipe-app into setup

touredri committed:
update readme file

Kwaku28 committed:
Update README.md

touredri committed:
fixed linters

touredri committed:
Merge branch 'setup' of https://github.com/touredri/Recipe-app into setup

Kwaku28 committed:
Update Gemfile.lock

Kwaku28 committed:
Fix rubocop error issues

Kwaku28 committed:
Remove unnecessary empty initialize method.

Kwaku28 committed:
Fix indentation

Kwaku28 committed:
Update ability.rb

touredri committed:
Merge pull request #12 from touredri/setup

Project Setup

Kwaku28 committed:
Generate models for users, foods, and recipes

Use scaffold to generate the models

touredri committed:
linters

touredri committed:
Update linters.yml

touredri committed:
Update linters.yml

touredri committed:
linters

touredri committed:
linters

touredri committed:
linters

touredri committed:
comment empty block

touredri committed:
Merge pull request #13 from touredri/database-setup

Generate models for users, foods, and recipes

touredri committed:
added food list

touredri committed:
Merge pull request #14 from touredri/food_list

added food list

Kwaku28 committed:
Generate migration to edit column data type

Kwaku28 committed:
Update CRUD method for recipe

Kwaku28 committed:
Set relationship between tables

Kwaku28 committed:
Update the corresponding view for recipie

Kwaku28 committed:
Update database.yml

Kwaku28 committed:
Fix rubocop issues

touredri committed:
Merge pull request #15 from touredri/add-recipe

Add recipe

Kwaku28 committed:
Delete files

Kwaku28 committed:
Define user abilities

Kwaku28 committed:
Add method to show public recipes

Kwaku28 committed:
Update CRUD method for adding ingredient

Kwaku28 committed:
Set up `recipe_foods` associations

Kwaku28 committed:
Set up the corresponding views

touredri committed:
Merge pull request #16 from touredri/view-recipe

View recipe

Kwaku28 committed:
Make changes to recipe

Kwaku28 committed:
Merge pull request #17 from touredri/view-recipe

Make changes to recipe

Kwaku28 committed:
Generate devise and add views

Kwaku28 committed:
Merge pull request #18 from touredri/setup-device

Generate devise and add views

touredri committed:
pubic recipe view added

touredri committed:
fixed linters

Kwaku28 committed:
Merge pull request #19 from touredri/public_recipe

pubic recipe view added

Kwaku28 committed:
Delete unwanted files

Kwaku28 committed:
Update app to add ingredient, toggle between public and private

Kwaku28 committed:
Merge branch 'dev' into add-ingredient

Kwaku28 committed:
Merge pull request #20 from touredri/add-ingredient

Add ingredient

touredri committed:
added shopping list

touredri committed:
css linters fixed

Kwaku28 committed:
Merge pull request #21 from touredri/chopping_list

added shopping list

Kwaku28 committed:
Delete edit.html.erb

Kwaku28 committed:
Add basic styling to page

Kwaku28 committed:
Delete comments and commented codes

Kwaku28 committed:
Create a navigation for the app

Kwaku28 committed:
Render navigation to pages

Kwaku28 committed:
Render navigation to pages

Kwaku28 committed:
Define class name for the forms

Kwaku28 committed:
Render navigation to page and refactor code

Kwaku28 committed:
Render navigation to page

Kwaku28 committed:
Add button to delete ingredient

Kwaku28 committed:
Merge branch 'dev' into navigation

Kwaku28 committed:
Update path to shopping list page

Kwaku28 committed:
Add basic styling to shopping list page

Kwaku28 committed:
Add basic styling to button

touredri committed:
Merge pull request #22 from touredri/navigation

Navigation

touredri committed:
test part started

touredri committed:
models test added

Kwaku28 committed:
Set up testing environment

Kwaku28 committed:
Create .rspec

Kwaku28 committed:
Place content in a `div` element

Kwaku28 committed:
Add basic styling to sign in page

Kwaku28 committed:
Add unit test to check the functionality

Kwaku28 committed:
Fix robocop issues

Kwaku28 committed:
Set condition to recipe index to hide recipes not created by the user

touredri committed:
added tst

touredri committed:
linters

touredri committed:
linters

Kwaku28 committed:
Merge pull request #23 from touredri/test

Test

touredri committed:
eager

touredri committed:
auth

aamirkhan2478 commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+class UsersController < ApplicationController
+  def shopping_list
+    @shopping_list = []
+    all_foods = Food.all
+    foods_in_recipes = Food.joins(:recipes).distinct
+    all_foods.each do |food|
+      @shopping_list << food unless foods_in_recipes.include?(food)
+    end
+    @total_item = @shopping_list.count
+    @total_price = @shopping_list.sum { |food| food.price * food.quantity }
+  end
```
- [ ] Please show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user) but for now I didn't see any data on the shopping list it's totally empty tables if the user add Ingredient food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user) so kindly fix this issue as well 
- [ ] Please count the total food items and the total price of the missing food and show it in the shopping list.

touredri replied:
@aamirkhan2478 can you explain us more the general shopping list

aamirkhan2478 commented on Gemfile:
```diff
@@ -0,0 +1,87 @@
+
+group :test do
+  # Use system testing [https://guides.rubyonrails.org/testing.html#system-testing]
+  gem 'capybara'
```
- [ ] Please add this `gem 'capybara'` inside the `group :development, :test do` and also configure it in your `spec/rails_helper.rb` file.
 - [ ] Please use integration tests by using capabara for now you didn't write any integration tests

touredri committed:
auth

aamirkhan2478 reviewed (changes requested):
Hi @Kwaku28 & @touredri , 

![](https://media2.giphy.com/media/t8YPf3ACKeouLzEIvr/giphy.gif?cid=ecf05e47k5uapdptg00nidbw3qaxx8iqgsusgaswci2c3uh1&rid=giphy.gif&ct=g)

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights
✔️ Good Effort
✔️ Your design is good.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

touredri commented on app/controllers/foods_controller.rb:
```diff
@@ -0,0 +1,59 @@
+    respond_to do |format|
+      if @food.save
+        format.html { redirect_to foods_url, notice: 'Food was successfully created.' }
+        # format.json { render :show, status: :created, location: @food }
+      else
+        format.html { render :new, status: :unprocessable_entity }
+        # format.json { render json: @food.errors, status: :unprocessable_entity }
```
@aamirkhan2478  ok just now

touredri committed:
auth

touredri committed:
auth

Kwaku28 committed:
Made changes based on review

Add Integration tests, Refactored shopping list view to display the correct contents, Add more tests.

Kwaku28 committed:
Fix rubocop issues

Kwaku28 committed:
Merge pull request #25 from touredri/hot-fix

Hot fix

tugsem commented on app/controllers/foods_controller.rb:
```diff
@@ -0,0 +1,59 @@
+
+  # GET /foods or /foods.json
+  def index
+    @foods = Food.all
```
- [ ] As the previous reviewer said, please render the foods belonging to the current user. You can use the 'current_user' method that comes with the devise.

rloterh replied:
well resolved! ✔️

tugsem commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class UsersController < ApplicationController
+  def shopping_list
+    @foods = Food.all
+    @recipe_foods = Food.joins(:recipes).distinct
+    @shopping_list = []
+    @recipe_foods.each do |recipe_food|
+      @shopping_list << recipe_food
+    end
+    @final_list = @foods.where(id: @shopping_list.map(&:id))
+
+    @price = @final_list.pluck(:price)
+    @quantity = @final_list.pluck(:quantity)
+    @total = @price.zip(@quantity).map { |x, y| x * y }.sum
```
- [ ] As requested by the previous reviewer, please fix the shopping list. Right now, the list shows the exact quantity of the item as an ingredient for a recipe. But the foods belong to a single user. Think of it as a stock. So I already have that ingredient as 500 grams. But recipe wants me to add 750gr. 

When the shopping list generated, user should see of the ingredient needed as 250 grams. The price should re-calculated. Kindly add this functionality.

rloterh replied:
well resolved! ✔️

tugsem commented on app/views/foods/index.html.erb:
```diff
@@ -0,0 +1,31 @@
+      </tr>
+    </thead>
+    <tbody>
+      <% @foods.each do |food| %>
+        <% if can?(:manage, food) %>
+          <%= render food %>
+        <% end %>
```
- [ ] As requested by the previous reviewer, you should only set @foods to the logged in user. So that here, you don't need to operate and consume time to load all of the foods.

rloterh replied:
well resolved! ✔️

tugsem commented:
## STATUS: REQUIRED CHANGES ♻️
Hello @Kwaku28 👋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are some issues that you still need to work on to go to the next project but you are almost there!

![ Alt Text](https://media.giphy.com/media/l2Je9zHYveK012EVi/giphy.gif)

 ## Highlights 🚀
✔️Good job on implementing the changes

 ## Required Changes ♻️
_Check the comments above the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

touredri committed:
linters

Kwaku28 committed:
Make changes according to review

Kwaku28 committed:
Merge pull request #26 from touredri/hot-fix

Hot fix

touredri committed:
Update users_controller.rb

touredri committed:
Update shopping_list.html.erb

rloterh commented:
@touredri 
Kindly fix linter errors!

Kwaku28 committed:
Make changes to Shopping list

Kwaku28 committed:
Merge branch 'dev' into hot-fix

Kwaku28 committed:
Fix errors

Kwaku28 committed:
Merge branch 'hot-fix' of https://github.com/touredri/Recipe-app into hot-fix

Kwaku28 committed:
Update users_controller.rb

Kwaku28 committed:
Merge pull request #27 from touredri/hot-fix

Make changes to Shopping list

Kwaku28 committed:
Make changes to shopping list

rloterh reviewed (approved):
## Status: Approved :heavy_check_mark:

Hi Drissa & Lawrence,  

Your project is complete! 🏁 Nice job! 💎💎 There is nothing else to say other than... it's time to merge it 🍾 Congratulations!  🎉💯🌟 

## Highlights 🎯 
🎲 Good job on introducing improvements to the app's functionality!
🎲 ReadMe looks good!

Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
